src/preamble.js

```diff
@@ -1223,4 +1223,13 @@ function runMemoryInitializer() {
 }
 #endif // MEM_INIT_IN_WASM == 0
 
+#if MAIN_MODULE && ASYNCIFY
+// With MAIN_MODULE + ASYNCIFY the normal method of placing stub functions in
+// wasmImports for as-yet-undefined symbols doesn't work since ASYNCIFY then
+// wraps these stub functions and we can't then replace them directly.  Instead
+// the stub functions call into `asyncifyStubs` which gets populated by the
+// dyanmic linker and symbols are loaded.
```

Comment:
Which functions end up here? I thought all of them, initially, but reading other code I am not sure.

Reply:
That is, one place seems to always read from `asyncifyStubs`, but another place checks if a thing is there or not, so I am confused.

Reply:
Only functions for which we create stubs need to appear in this map.   Its the same set of functions that get the `.stub` attribute set up them.  These are functions which are required to instantiate the main module, but which are provided later, by a side module.

Without ASYNCIFY we simply replace the existing `.stub` function wasmImports once the symbols gets resolved.
With ASYNCIFY this appraoch doesn't work because the thing in the wasmImports dict might be an asyncify wrapper around the `.stub` we created.   In this case we replace the entry in `asyncifyStubs` which is called by the stub function.

Reply:
There are two places the `asyncifyStubs` are used.

1. In the stub functions themselves.   In debug builds we assert that that right entry is present.  In release builds we will just trap if its not present.
2. In the dynamic linker: When a new symbol is loaded it replaces the entry in asyncifyStubs, but only if that key already exists in the dictionary.